File: src/service/data/IAirQualityDataService.java
package service.data;

import model.QualityIndex;
import model.Sensor;
import model.SensorData;
import model.Station;

import java.util.List;
import java.util.concurrent.CompletableFuture;

/**
 * Class that is responsible for managing data from different sources
 */
public interface IAirQualityDataService {

    /**
     * Method that download stations from sources
     *
     * @return list of stations
     */
    CompletableFuture<List<Station>> getStations();
    /**
     * Method that download sensor for station from sources
     *
     * @return list of sensor
     */
    CompletableFuture<List<Sensor>> getSensors(Long stationId);
    /**
     * Method that download sensor data for sensor from sources
     *
     * @return list of sensordata
     */
    CompletableFuture<List<SensorData>> getSensorData(Sensor sensor);
    /**
     * Method that download indices for given station from sources
     *
     * @return list of indices
     */
    CompletableFuture<List<QualityIndex>> getIndexes(Long stationId);

}

File: src/service/data/IAirQualityDataService.kt
package service.data

import kotlinx.coroutines.Deferred
import model.QualityIndex
import model.Sensor
import model.SensorData
import model.Station

/**
 * Class that is responsible for managing data from different sources
 */
interface IAirQualityDataService {

    /**
     * Method that download stations from sources
     *
     * @return list of stations
     */
    fun getStations(): Deferred<List<Station>>

    /**
     * Method that download sensor for station from sources
     *
     * @return list of sensor
     */
    fun getSensors(stationId: Long): Deferred<List<Sensor>>

    /**
     * Method that download sensor data for sensor from sources
     *
     * @return list of sensordata
     */
    fun getSensorData(sensor: Sensor): Deferred<List<SensorData>>

    /**
     * Method that download indices for given station from sources
     *
     * @return list of indices
     */
    fun getIndexes(stationId: Long): Deferred<List<QualityIndex>>

}
